File: rectangle.cpp
#include "rectangle.h"

//constructor
Rectangle::Rectangle() {
    high_left_x_ = 0;
    high_left_y_ = 0;
    low_right_x_ = 0;
    low_right_y_ = 0;
}

// destructor
Rectangle::~Rectangle() {
    high_left_x_ = 0;
    high_left_y_ = 0;
    low_right_x_ = 0;
    low_right_y_ = 0;
}

// input data for rectangle
void Rectangle::Input(ifstream &ifst) {
    ifst >> high_left_x_ >> high_left_y_ >> low_right_x_
    >> low_right_y_;
    is_valid_ = ValidateRectangle();
}

// check input data
bool Rectangle::ValidateRectangle() {
    if(high_left_x_ < low_right_x_ && high_left_y_ > low_right_y_) {
        return true;
    }
    return false;
}

// get random data for rectangle
void Rectangle::InputRandom() {
    high_left_x_ = Random();
    high_left_y_ = Random();
    low_right_x_ = Random();
    low_right_y_ = Random();
    is_valid_ = ValidateRectangle();
}

// output data of rectangle
void Rectangle::Output(ofstream &ofst) {
    ofst << " Rectangle: highLeftX = " << high_left_x_ 
    << ", highLeftY = " << high_left_y_
    << ", lowRightX = " << low_right_x_
    << ", lowRightY = " << low_right_y_
    << ". Perimeter = " << Perimeter() << "\n";
    if(is_valid_ == false) {
        ofst << "Earlier Rectangle was not valid \n";
    }
}

// get perimeter of rectangle
double Rectangle::Perimeter() {
    return 2 * (std::abs(low_right_x_ - high_left_x_) +
    std::abs(high_left_y_ - low_right_y_));
}

File: rectangle.py
from shape import Shape


class Rectangle(Shape):
    def __init__(self, color, line):
        Shape.__init__(self, color)
        self.is_valid = True
        self.input(line)

    def input(self, line):
        Shape.set(self, line, 2)

    def output(self, file):
        out_str = (self.color + " Rectangle: " +
                                " hightLeftX = " + str(self.nodes[0].X) +
                                " hightLeftY = " + str(self.nodes[0].Y) +
                                " lowLeftX = " + str(self.nodes[0].X) +
                                " lowRightX = " + str(self.nodes[1].Y) +
                                " Perimeter = " + str(self.get_perimeter()) + "\n")
        file.write(out_str)

        if not self.is_valid:
            out_str = "This input was invalid\n"
            file.write(out_str)

    def get_perimeter(self):
        horizontal = self.nodes[0].X - self.nodes[1].X
        vertical = self.nodes[0].Y - self.nodes[1].Y
        if horizontal < 0 or vertical < 0:
            self.is_valid = False
        return 2 * (abs(horizontal) + abs(vertical))
